src/rgw/rgw_common.cc

```diff
@@ -1242,6 +1242,17 @@ Effect evaluate_iam_policies(
   return Effect::Pass;
 }
 
+bool verify_bucket_expected_ownership(req_state *const s)
+{
+  if (s->info.env->get("HTTP_X_AMZ_EXPECTED_BUCKET_OWNER") != nullptr){
+    auto expected_bucket_owner = s->info.env->get("HTTP_X_AMZ_EXPECTED_BUCKET_OWNER");
+    if (expected_bucket_owner != to_string(s->bucket->get_owner())){
```

Comment:
`get_owner()` returns `rgw_owner`, which is either a `rgw_user` or `rgw_account_id`. the latter corresponds to the AWS behavior documented under [Restrictions and limitations](https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucket-owner-condition.html#bucket-owner-condition-restrictions-limitations):

> The value of the bucket owner condition parameter must be an AWS account ID (12-digit numeric value).

and the string-based comparison for `rgw_account_id` will work fine :+1:

but for `rgw_user` there's an additional field `rgw_user::tenant` that `rgw_user::to_str()` renders as `"tenant$id"`, and s3 clients will expect the string `"id"` to work (whether or not the bucket lives in a tenant)